File: src/main/java/fit/wenchao/db/JavaSourceFile.java
package fit.wenchao.db;

import lombok.var;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

public class JavaSourceFile {
    JavaClassName javaClassName;
    String srcFileName;
    String srcCode;

    String packageName;

    public static JavaSourceFile ofMysqlModel(Table table, JavaPackage javaPackage) {
        JavaSourceFile javaSourceFile = new JavaSourceFile();
        StringBuilder codeBuilder = new StringBuilder();

        JavaClassName javaClassName = JavaClassName.fromLowerUnderScore(javaPackage, table.getName(), "PO");
        javaSourceFile.srcFileName = javaClassName.toJavaSrcFileName();
        javaSourceFile.javaClassName = javaClassName;
        javaSourceFile.packageName = javaPackage.getDotSplitName();
        //JavaClassName javaClassName = JavaClassName.fromLowerUnderScore("user_action");

        JavaCode javaCode = JavaCode
                .writing()
                .packagel(javaPackage.getDotSplitName())
                .importl("lombok.Data")
                .atl("Data")
                .publicl().classl(javaClassName.getName())
                .blockl(jc -> {
                    for (TableAttr tableAttr : table) {
                        if (tableAttr.getType().equalsIgnoreCase("varchar")) {
                            jc.write("String ");
                            //codeBuilder.append("String ");
                        }
                        if (tableAttr.getType().equalsIgnoreCase("int")) {
                            jc.write("int ");
                            //codeBuilder.append("int ");
                        }
                        JavaVarName javaVarName = JavaVarName.fromUnderScore(tableAttr.getName());
                        jc.write(javaVarName.getName() + ";\n");
                        //codeBuilder.append(javaVarName.getName() + ";\n");
                    }
                });

        javaSourceFile.srcCode = javaCode.toString();
        return javaSourceFile;
    }

    public static JavaSourceFile ofMybatisMapper(Table table, JavaPackage javaPackage, JavaSourceFile poSourceFile) {
        JavaSourceFile javaSourceFile = new JavaSourceFile();
        StringBuilder codeBuilder = new StringBuilder();

        JavaClassName poClassName = poSourceFile.getJavaClassName();

        JavaClassName mapperClassName = JavaClassName.fromLowerUnderScore(javaPackage, table.getName(), "Mapper");
        //poSourceFile.
        javaSourceFile.srcFileName = mapperClassName.toJavaSrcFileName();
        javaSourceFile.javaClassName = mapperClassName;
        javaSourceFile.packageName = javaPackage.getDotSplitName();
        //JavaClassName javaClassName = JavaClassName.fromLowerUnderScore("user_action");


        JavaCode javaCode =
                JavaCode.writing()
                        .packagel(javaPackage.getDotSplitName())
                        .importl(poClassName.getFullName())
                        .importl("org.apache.ibatis.annotations.Mapper")
                        .importl("com.baomidou.mybatisplus.core.mapper.BaseMapper")
                        .atl("Mapper")
                        .publicl().interfacel(mapperClassName.getName()).extendsl("BaseMapper<" + poClassName.getName() + ">")
                        .blockl();

        javaSourceFile.srcCode = javaCode.toString();
        return javaSourceFile;
    }

    public static JavaSourceFile ofServiceImpl(Table table, JavaPackage javaPackage, JavaSourceFile serviceSourceFile) {
        JavaSourceFile javaSourceFile = new JavaSourceFile();
        StringBuilder codeBuilder = new StringBuilder();

        JavaClassName serviceClassName = serviceSourceFile.getJavaClassName();

        JavaClassName serviceImplClassName = JavaClassName.fromLowerUnderScore(javaPackage, table.getName(), "ServiceImpl");
        //poSourceFile.
        javaSourceFile.srcFileName = serviceImplClassName.toJavaSrcFileName();
        javaSourceFile.javaClassName = serviceImplClassName;
        javaSourceFile.packageName = javaPackage.getDotSplitName();
        //JavaClassName javaClassName = JavaClassName.fromLowerUnderScore("user_action");
        //TODO
        JavaCode javaCode =
                JavaCode.writing()
                        .packagel(javaPackage.getDotSplitName())
                        .importl(serviceClassName.getFullName())
                        .importl("org.apache.ibatis.annotations.Mapper")
                        .importl("com.baomidou.mybatisplus.core.mapper.BaseMapper")
                        .atl("Mapper")
                        .publicl().interfacel(serviceImplClassName.getName()).extendsl("BaseMapper<" + serviceClassName.getName() + ">")
                        .blockl();

        javaSourceFile.srcCode = javaCode.toString();
        return javaSourceFile;
        return null;
    }


    private JavaClassName getJavaClassName() {
        return this.javaClassName;
    }

    public static JavaSourceFile ofDao(Table table, JavaPackage javaPackage) {
        JavaSourceFile javaSourceFile = new JavaSourceFile();
        StringBuilder codeBuilder = new StringBuilder();

        JavaClassName daoClassName = JavaClassName.fromLowerUnderScore(javaPackage, table.getName(), "Dao");
        javaSourceFile.srcFileName = daoClassName.toJavaSrcFileName();
        javaSourceFile.javaClassName = daoClassName;
        javaSourceFile.packageName = javaPackage.getDotSplitName();
        //JavaClassName javaClassName = JavaClassName.fromLowerUnderScore("user_action");

        JavaCode javaCode = JavaCode
                .writing()
                .packagel(javaPackage.getDotSplitName())
                .publicl().interfacel(daoClassName.getName())
                .blockl();

        javaSourceFile.srcCode = javaCode.toString();
        return javaSourceFile;
    }

    public static JavaSourceFile ofDaoImpl(Table table, JavaPackage javaPackage, JavaSourceFile daoSourceFile) {
        JavaSourceFile javaSourceFile = new JavaSourceFile();
        StringBuilder codeBuilder = new StringBuilder();

        JavaClassName daoImplClassName = JavaClassName.fromLowerUnderScore(javaPackage, table.getName(), "DaoImpl");
        javaSourceFile.srcFileName = daoImplClassName.toJavaSrcFileName();
        javaSourceFile.javaClassName = daoImplClassName;
        javaSourceFile.packageName = javaPackage.getDotSplitName();
        //JavaClassName javaClassName = JavaClassName.fromLowerUnderScore("user_action");

        JavaCode javaCode =
                JavaCode.writing()
                        .packagel(javaPackage.getDotSplitName())
                        .importl("org.springframework.stereotype.Repository")
                        .importl(daoSourceFile.getJavaClassName().getFullName())
                        .atl("Repository")
                        .publicl().classl(daoImplClassName.getName()).implementsl(daoSourceFile.getJavaClassName().getName())
                        .blockl();

        javaSourceFile.srcCode = javaCode.toString();
        return javaSourceFile;
    }

    public void put2Package(JavaPackage javaPackage) {
        File javaPackageFile = javaPackage.getFile();
        Path javaSourceFilePath = javaPackageFile.toPath().resolve(Paths.get(this.srcFileName));
        try (
                OutputStream out = Files.newOutputStream(javaSourceFilePath);
                ByteArrayInputStream in = new ByteArrayInputStream(this.srcCode.getBytes(StandardCharsets.UTF_8));
        ) {
            byte[] buffer = new byte[1024 * 1024];
            int len;
            while ((len = in.read(buffer)) != -1) {
                out.write(buffer, 0, len);
            }
            out.flush();
        }
        catch (IOException e) {
            throw new RuntimeException(e);
        }
    }


    public static JavaSourceFile ofService(Table table, JavaPackage javaPackage) {
        JavaSourceFile javaSourceFile = new JavaSourceFile();


        JavaClassName serviceClassName = JavaClassName.fromLowerUnderScore(javaPackage, table.getName(), "Service");
        //poSourceFile.
        javaSourceFile.srcFileName = serviceClassName.toJavaSrcFileName();
        javaSourceFile.javaClassName = serviceClassName;
        javaSourceFile.packageName = javaPackage.getDotSplitName();
        //JavaClassName javaClassName = JavaClassName.fromLowerUnderScore("user_action");

        JavaCode javaCode =
                JavaCode.writing()
                        .packagel(javaPackage.getDotSplitName())
                        .publicl().interfacel(serviceClassName.getName())
                        .blockl();

        javaSourceFile.srcCode = javaCode.toString();
        return javaSourceFile;
    }

    public boolean existsIn(JavaPackage javaPackage) {
        var srcNames = javaPackage.listSourceFileNames();
        return srcNames.contains(this.srcFileName);
    }
}

File: src/main/kotlin/fit/wenchao/db/JavaSourceFile.kt
package fit.wenchao.db

import fit.wenchao.db.JavaClassName.Companion.fromLowerUnderScore
import fit.wenchao.db.codeWriter.JavaCodeWriter
import fit.wenchao.db.codeWriter.KotlinCodeWriter
import fit.wenchao.db.codeWriter.javaWriter
import fit.wenchao.db.codeWriter.kotlinWriter
import fit.wenchao.db.constants.Lang
import fit.wenchao.db.logger.Log
import java.io.ByteArrayInputStream
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Paths
import java.util.*

class JavaSourceFile: Log {

    var javaPackage: JavaPackage
    var javaClassName: JavaClassName
    var srcFileName: String
    var srcCode: String
    var packageName: String

    constructor(clazzName: JavaClassName, language: Lang, srcCode:String)  {
        this.srcFileName = clazzName.toSrcFileName(language)
        this.javaClassName = clazzName
        this.javaPackage = clazzName.javaPackage
        this.packageName = clazzName.javaPackage.dotSplitName
        this.srcCode = srcCode
    }

    fun put2Package(javaPackage: JavaPackage) {
        val javaPackageFile = javaPackage.file
        val javaSourceFilePath = javaPackageFile.toPath().resolve(Paths.get(srcFileName))

        try {
            Files.newOutputStream(javaSourceFilePath).use { out ->
                ByteArrayInputStream(srcCode.toByteArray(StandardCharsets.UTF_8)).use { `in` ->
                    val buffer = ByteArray(1024 * 1024)
                    var len: Int
                    while (`in`.read(buffer).also { len = it } != -1) {
                        out.write(buffer, 0, len)
                    }
                    out.flush()
                }
            }
        } catch (e: IOException) {
            throw RuntimeException(e)
        }

    }


    fun put2Package() {
        val javaPackageFile = javaPackage.file
        val javaSourceFilePath = javaPackageFile.toPath().resolve(Paths.get(srcFileName))
        try {
            Files.newOutputStream(javaSourceFilePath).use { out ->
                ByteArrayInputStream(srcCode.toByteArray(StandardCharsets.UTF_8)).use { `in` ->
                    val buffer = ByteArray(1024 * 1024)
                    var len: Int
                    while (`in`.read(buffer).also { len = it } != -1) {
                        out.write(buffer, 0, len)
                    }
                    out.flush()
                }
            }
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    fun notExistsIn(javaPackage: JavaPackage): Boolean {
        val srcNames = javaPackage.listSourceFileNames()
        return !srcNames.contains(srcFileName)
    }

    fun notExists(): Boolean {
        val srcNames = javaPackage.listSourceFileNames()
        return !srcNames.contains(srcFileName)
    }

}

fun ofMysqlModel(table: Table, javaPackage: JavaPackage, lang: Lang): JavaSourceFile {


    table.name ?: throw RuntimeException()


    val javaClassName = fromLowerUnderScore(javaPackage, table.name!!, "PO")


    // if kotlin
    var code = if (lang == Lang.KOTLIN) {
        kotlinWriter()
            .packagel(javaPackage.dotSplitName)
            .importl("com.baomidou.mybatisplus.annotation.TableName")
            .importl("com.baomidou.mybatisplus.annotation.TableId")
            .importl("com.baomidou.mybatisplus.annotation.IdType")
            .importl("java.io.Serializable")
            .atl("TableName(\"`" + table.name + "`\")")
            .datal().classl(javaClassName.name).primaryConstructor { jc: KotlinCodeWriter ->
                val size = table.attrs.size
                var count = 0
                for (tableAttr in table) {
                    count++
                    if (tableAttr.isPri) {
                        //@TableId(value = "id", type = IdType.AUTO)
                        jc.atl("TableId(value=\"" + tableAttr.name + "\", type=IdType.AUTO)")
                    }
                    jc.write("var ")
                    val javaVarName = fromUnderScore(tableAttr.name)
                    jc.write(javaVarName.name + ": ")
                    if (tableAttr.type.equals("int", ignoreCase = true)) {
                        jc.write("Int ")
                        // codeBuilder.append("int ");
                    }
                    if (tableAttr.type.equals("bigint", ignoreCase = true)) {
                        jc.write("Long ")
                        // codeBuilder.append("int ");
                    }
                    if (tableAttr.type.lowercase(Locale.getDefault()).contains("text")
                        || tableAttr.type.lowercase(Locale.getDefault()).contains("json")
                        || tableAttr.type.equals("varchar", ignoreCase = true)
                    ) {
                        jc.write("String ")
                    }
                    if (count == size) {
                        jc.write("?")
                    } else {
                        jc.write("?,")
                    }
                    jc.write("\n")
                }
                null
            }.extendsFirstl("Serializable")
            .toString()
    } else {
        javaWriter()
            .packagel(javaPackage.dotSplitName)
            .importl("com.baomidou.mybatisplus.annotation.TableName")
            .importl("com.baomidou.mybatisplus.annotation.TableId")
            .importl("com.baomidou.mybatisplus.annotation.IdType")
            .importl("java.io.Serializable")
            .importl("lombok.Data")
            .importl("lombok.AllArgsConstructor")
            .importl("lombok.Builder")
            .importl("lombok.Data")
            .importl("lombok.NoArgsConstructor")
            .importl("lombok.experimental.Accessors")
            .atl("Data")
            .atl("AllArgsConstructor")
            .atl("NoArgsConstructor")
            .atl("Builder")
            .atl("Accessors(chain = true)")
            .atl("TableName(\"`" + table.name + "`\")")
            .publicl().classl(javaClassName.name).implementsl("Serializable")
            .blockl { jc: JavaCodeWriter ->
                for (tableAttr in table) {
                    if (tableAttr.isPri) {
                        //@TableId(value = "id", type = IdType.AUTO)
                        jc.atl("TableId(value=\"" + tableAttr.name + "\", type=IdType.AUTO)")
                    }
                    if (tableAttr.type.equals("varchar", ignoreCase = true)) {
                        jc.write("String ")
                        // codeBuilder.append("String ");
                    }
                    if (tableAttr.type.equals("int", ignoreCase = true)) {
                        jc.write("int ")
                        // codeBuilder.append("int ");
                    }
                    val javaVarName = fromUnderScore(tableAttr.name)
                    jc.write(javaVarName.name + ";\n")
                    // codeBuilder.append(javaVarName.name + ";\n");
                }
            }.toString()
    }
    val javaSourceFile = JavaSourceFile(javaClassName, lang, code)
    return javaSourceFile
}

fun ofMybatisMapper(
    table: Table,
    javaPackage: JavaPackage,
    poSourceFile: JavaSourceFile,
    lang: Lang
): JavaSourceFile {
    val poClassName = poSourceFile.javaClassName
    val mapperClassName = fromLowerUnderScore(javaPackage, table.name, "Mapper")
    var code: String = if (lang == Lang.KOTLIN) {
        val javaCode = kotlinWriter()
            .packagel(javaPackage.dotSplitName)
            .importl(poClassName!!.getFullName())
            .importl("org.apache.ibatis.annotations.Mapper")
            .importl("com.baomidou.mybatisplus.core.mapper.BaseMapper")
            .atl("Mapper")
            .interfacel(mapperClassName.name).extendsFirstl("BaseMapper<" + poClassName.name + ">")
            .blockl()
        javaCode.toString()
    } else {
        val javaCode = javaWriter()
            .packagel(javaPackage.dotSplitName)
            .importl(poClassName!!.getFullName())
            .importl("org.apache.ibatis.annotations.Mapper")
            .importl("com.baomidou.mybatisplus.core.mapper.BaseMapper")
            .atl("Mapper")
            .publicl().interfacel(mapperClassName.name).extendsl("BaseMapper<" + poClassName.name + ">")
            .blockl()
        javaCode.toString()
    }
    val javaSourceFile = JavaSourceFile(mapperClassName, lang, code)
    return javaSourceFile
}

fun ofServiceImpl(
    table: Table,
    javaPackage: JavaPackage,
    serviceSourceFile: JavaSourceFile,
    lang: Lang
): JavaSourceFile {
    val serviceClassName = serviceSourceFile.javaClassName
    val serviceImplClassName = fromLowerUnderScore(
        javaPackage,
        table.name,
        "ServiceImpl"
    )
    var code: String = if (lang == Lang.KOTLIN) {
        val javaCode = kotlinWriter()
            .packagel(javaPackage.dotSplitName)
            .importl(serviceClassName!!.getFullName())
            .importl("org.springframework.stereotype.Service")
            .atl("Service")
            .classl(serviceImplClassName.name).extendsFirstl(serviceSourceFile.javaClassName!!.name)
            .blockl()
        javaCode.toString()
    } else {
        val javaCode = javaWriter()
            .packagel(javaPackage.dotSplitName)
            .importl(serviceClassName!!.getFullName())
            .importl("lombok.extern.slf4j.Slf4j")
            .importl("org.springframework.stereotype.Service")
            .atl("Service")
            .atl("Slf4j")
            .publicl().classl(serviceImplClassName.name).implementsl(serviceSourceFile.javaClassName!!.name)
            .blockl()
        javaCode.toString()
    }
    val javaSourceFile = JavaSourceFile(serviceImplClassName, lang, code)
    return javaSourceFile
}

fun ofDao(
    table: Table,
    javaPackage: JavaPackage,
    poJavaSourceFile: JavaSourceFile,
    lang: Lang
): JavaSourceFile {
    val daoClassName = fromLowerUnderScore(javaPackage, table.name, "Dao")
    var code: String = if (lang == Lang.KOTLIN) {
        val javaCode = kotlinWriter()
            .packagel(javaPackage.dotSplitName)
            .importl(poJavaSourceFile.javaClassName!!.getFullName())
            .importl("com.baomidou.mybatisplus.extension.service.IService")
            .interfacel(daoClassName.name)
            .extendsFirstl("IService<" + poJavaSourceFile.javaClassName!!.name + ">")
            .blockl()
        javaCode.toString()
    } else {
        val javaCode = javaWriter()
            .packagel(javaPackage.dotSplitName)
            .importl(poJavaSourceFile.javaClassName!!.getFullName())
            .importl("com.baomidou.mybatisplus.extension.service.IService")
            .publicl().interfacel(daoClassName.name)
            .extendsl("IService<" + poJavaSourceFile.javaClassName!!.name + ">")
            .blockl()
        javaCode.toString()
    }
    val javaSourceFile = JavaSourceFile(daoClassName, lang, code)
    return javaSourceFile
}

fun ofDaoImpl(
    table: Table,
    javaPackage: JavaPackage,
    mapperSourceFile: JavaSourceFile,
    poSourceFile: JavaSourceFile,
    daoSourceFile: JavaSourceFile,
    lang: Lang
): JavaSourceFile {
    val daoImplClassName = fromLowerUnderScore(javaPackage, table.name, "DaoImpl")
    var code: String = if (lang == Lang.KOTLIN) {
        val javaCode = kotlinWriter()
            .packagel(javaPackage.dotSplitName)
            .importl("org.springframework.stereotype.Repository")
            .importl("com.baomidou.mybatisplus.extension.service.impl.ServiceImpl")
            .importl(poSourceFile.javaClassName!!.getFullName())
            .importl(mapperSourceFile.javaClassName!!.getFullName())
            .importl(daoSourceFile.javaClassName!!.getFullName())
            .atl("Repository")
            .classl(daoImplClassName.name)
            .extendsFirstl("ServiceImpl<" + mapperSourceFile.javaClassName!!.name + "," + poSourceFile.javaClassName!!.name + ">()")
            .extendsThenl(daoSourceFile.javaClassName!!.name)
            .blockl()
        javaCode.toString()
    } else {
        val javaCode = javaWriter()
            .packagel(javaPackage.dotSplitName)
            .importl("org.springframework.stereotype.Repository")
            .importl("com.baomidou.mybatisplus.extension.service.impl.ServiceImpl")
            .importl(poSourceFile.javaClassName!!.getFullName())
            .importl(mapperSourceFile.javaClassName!!.getFullName())
            .importl(daoSourceFile.javaClassName!!.getFullName())
            .atl("Repository")
            .publicl().classl(daoImplClassName.name)
            .extendsl("ServiceImpl<" + mapperSourceFile.javaClassName!!.name + "," + poSourceFile.javaClassName!!.name + ">")
            .implementsl(
                daoSourceFile.javaClassName!!.name
            )
            .blockl()
        javaCode.toString()
    }
    val javaSourceFile = JavaSourceFile(daoImplClassName, lang, code)
    return javaSourceFile
}

fun ofService(table: Table, javaPackage: JavaPackage, lang: Lang): JavaSourceFile {
    val serviceClassName = fromLowerUnderScore(javaPackage, table.name, "Service")
    var code: String = if (lang == Lang.KOTLIN) {
        val javaCode = kotlinWriter()
            .packagel(javaPackage.dotSplitName)
            .interfacel(serviceClassName.name)
            .blockl()
        javaCode.toString()
    } else {
        val javaCode = javaWriter()
            .packagel(javaPackage.dotSplitName)
            .publicl().interfacel(serviceClassName.name)
            .blockl()
        javaCode.toString()
    }
    val javaSourceFile = JavaSourceFile(serviceClassName, lang, code)
    return javaSourceFile
}
